lib/Sema/CSDiagnostics.cpp

```diff
@@ -984,9 +984,38 @@ bool MissingExplicitConversionFailure::diagnoseAsError() {
   return true;
 }
 
-bool MemberAccessOnOptionalBaseFailure::diagnoseAsError() {
+Type MemberAccessOnOptionalBaseFailure::getMemberBaseType() const {
+  if (getLocator()->isForKeyPathComponent()) {
+    if (auto *memberBaseTypeVar = MemberBaseType->getAs<TypeVariableType>()) {
+      return getSolution().getFixedType(memberBaseTypeVar);
+    }
+    return MemberBaseType;
+  }
+  return getType(getAnchor());
+}
+
+SourceRange MemberAccessOnOptionalBaseFailure::getMemberBaseSourceRange() const {
```

Comment:
I think it would be better if this was an override of `FailureDiagnostic::getSourceRange`.